std-bits/table/src/main/java/org/enso/table/data/column/builder/DoubleBuilder.java

```diff
@@ -49,13 +51,47 @@ public StorageType getType() {
     return FloatType.FLOAT_64;
   }
 
+  /**
+   * Converts the provided LongBuilder to a DoubleBuilder.
+   *
+   * <p>The original LongBuilder becomes invalidated after this operation and should no longer be
+   * used.
+   */
+  static DoubleBuilder retypeFromLongBuilder(LongBuilder longBuilder) {
+    int currentSize = longBuilder.currentSize;
+    DoubleBuilder newBuilder = new DoubleBuilder(longBuilder.isMissing, longBuilder.data, currentSize);
+
+    // Invalidate the old builder.
+    longBuilder.data = null;
+    longBuilder.isMissing = null;
+    longBuilder.currentSize = -1;
+
+    // Translate the data in-place to avoid unnecessary allocations.
+    for (int i = 0; i < currentSize; i++) {
+      if (!newBuilder.isMissing.get(i)) {
+        long currentIntegerValue = newBuilder.data[i];
+        double convertedFloatValue = newBuilder.convertIntegerToDouble(currentIntegerValue);
+        newBuilder.data[i] = Double.doubleToRawLongBits(convertedFloatValue);
+      }
+    }
+
+    return newBuilder;
+  }
+
   @Override
   public void appendNoGrow(Object o) {
     if (o == null) {
       isMissing.set(currentSize++);
-    } else {
+    } else if (NumericConverter.isDecimalLike(o)){
       double value = NumericConverter.coerceToDouble(o);
       data[currentSize++] = Double.doubleToRawLongBits(value);
+    } else if (NumericConverter.isCoercibleToLong(o)) {
+      long value = NumericConverter.coerceToLong(o);
+      double converted = convertIntegerToDouble(value);
+      data[currentSize++] = Double.doubleToRawLongBits(converted);
+    } else {
+      throw new IllegalStateException("Unexpected value type when appending to a DoubleBuilder: " + o.getClass().getCanonicalName() + "." +
+          " This is a bug in the Table library.");
```

Comment:
We check decimal and integer cases separately, to avoid implicit rounding-loss-of-precision and handle the loss of precision explicitly with `convertIntegerToDouble`.